### PR TITLE
Add IO lecture

### PR DESCRIPTION
**Pull request by Alex-deVis**
Demo for the IO lecture.

### CODE REVIEW

**Comment by razvand** [comment]
@Alex-deVis , please add `## Section Name`  as individual slides to separate sections of the chapter. This will also get rid of the linter warning for Markdown.

See [the slides for the "Software Stack" chapter](https://github.com/open-education-hub/operating-systems/pull/13/files).

File: content/chapters/io/lecture/.gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+/_site
+/slides.md
+notes.md
```
Comment (razvand) [inline content/chapters/io/lecture/.gitignore:3]: Add newline.

File: content/chapters/io/lecture/slides/intro.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+----
+
+#### Interactive Software is Good Software
+
+![interactive software](./media/interactive-software.png)
+
+----
+
+#### IO - The Need
```
Comment (razvand) [inline content/chapters/io/lecture/slides/intro.md:22]: I would replace this with an image showing a process doing computing. And then adding items showing input / output highlighting interactivity.

File: content/chapters/io/lecture/slides/intro.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+- The more data we provide to our application, the more relevant its result becomes
+- Interactivity
+
+----
+
+#### What We Want
+
+- Means for our application to communicate
+- Compatibility with various devices
+- Responsiveness (Non-blocking I/O)
```
Comment (razvand) [inline content/chapters/io/lecture/slides/intro.md:34]: We also want "Performance" - fast I/O interfaces (fast transfer, I/O multiplexing)

File: content/chapters/io/lecture/slides/linux-files.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
+## Linux Files
```
Comment (razvand) [inline content/chapters/io/lecture/slides/linux-files.md:1]: This seems to sudden. I suggest you first present the I/O interface, before moving to I/O devices.

I would first show how you would use the I/O before showing the lower parts.

File: content/chapters/io/lecture/slides/linux-files.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+---
+
+### I/O Devices
+
+- A hardware component that communicates with our application through bytes
+- Various devices fit this description:
```
Comment (razvand) [inline content/chapters/io/lecture/slides/linux-files.md:8]: replace with pictures

File: content/chapters/io/lecture/slides/linux-files.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+  - hard disk
+  - sensors
+
+How does the OS operates them?
+
+----
+
+### Working with I/O Devices
+
+- Linux abstracts an I/O device as a **special file**
```
Comment (razvand) [inline content/chapters/io/lecture/slides/linux-files.md:21]: Replace with figure.

File: content/chapters/io/lecture/slides/linux-files.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+- Going a step further, we use files to represent and organize persistent data
+  - **regular files** store data
+  - **directories** construct a file hierarchy
+  - **symbolic links** referrence other files (shortcuts)
+- File hierarchies dramatically reduce search times
+
+----
+
+### Interprocess Communication
```
Comment (razvand) [inline content/chapters/io/lecture/slides/linux-files.md:63]: This has to be a separate section. Mentioned the two types of endpoints. Those endpoints backed by storage / devices (file interface) and those endpoints pointing to another process (socket interface).

**Review by razvand — CHANGES_REQUESTED** [review]
I made some comments. It's a good start. It can be improved by adding more images and connecting the dots better. Some parts seem pretty "sudden" story-wise.

File: content/chapters/io/lecture/slides/linux-files.md (hunk 4 above)
Comment (Alex-deVis) [reply content/chapters/io/lecture/slides/linux-files.md]: I went with your approach - explaining the interface without going into detail about `file descriptors` and `devices` feels more natural.

**Review by razvand — APPROVED** [review]
Looks good, @Alex-deVis . I will approve and merge it like this. Then you will create a separate PR for the 2nd and 3rd part of the I/O lecture.

**Pull request merged**